transbot/models/transbot_project.py

```diff
@@ -0,0 +1,542 @@
+                                        'resource': resource['slug'],
+                                        'language': lang['code'],
+                                        'branch': branch.id})
+                # Force an update of the last commit of the branch
+                gh_branch = gh_repo.get_branch(branch.name)
+                branch.last_commit_sha = gh_branch.commit.sha
+
+    @api.multi
+    def check_transifex_updates_cron(self):
+        try:
+            return self.check_transifex_updates
+        except Exception as e:
+            self.env['transbot.log'].log_message(
+                _("Error executing Transifex update.\nDetails: %s" %
+                  e.message), log_type='error')
+
+    def check_and_touch_resource(self, branch, resource, last_update):
+        """An attempt of optimizing the query caching last check date.
+        It can be used for now because there's no info on last update on a
+        translation.
+        :param branch:
+        :param resource:
+        :param last_update:
+        :return: True if the resource has been updated
+        """
+        res_model = self.env['transbot.project.branch.resource']
+        res = res_model.search([('branch', '=', branch.id),
+                                ('resource', '=', resource)])
+        if res:
+            result = (fields.Datetime.from_string(res.last_update) <
+                      datetime.strptime(last_update, '%Y-%m-%dT%H:%M:%S.%f'))
+            res.last_update = fields.Datetime.now()
+            return result
+        else:
+            res_model.create({'resource': resource,
+                              'branch': branch.id})
+            return True
+
+    def _add_or_replace_github_translation(
+            self, gh_repo, gh_branch, project_slug, resource_slug, lang_code,
+            project):
+        if lang_code == 'en':
+            # We ignore english, because it's already in the template
+            return
+        tx = self.env['transbot.config.settings'].get_transifex_connection()
+        # Get file
+        po_file = tx.get('/api/2/project/%s/resource/%s/translation/%s' %
+                         (project_slug, resource_slug, lang_code))
+        # Check if file exists already in Github
+        path = '/%s/i18n/%s.po' % (resource_slug, lang_code)
+        try:
+            # Check if file exists
+            gh_file = gh_repo.get_contents(path, gh_branch.name)
+            sha = gh_file.sha
+        except UnknownObjectException:
+            sha = None
+        gh_repo.put_file_contents(
+            path, "Transbot '%s' translation automatic update for '%s'" %
+            (lang_code, resource_slug), po_file['content'], sha,
+            gh_branch.name)
+        self.env['transbot.log'].log_message(
+            _("Language '%s' for module '%s' updated on Github.") %
+            (lang_code, resource_slug), project=project.id)
+        return sha
+
+
+class GithubProjectBranch(models.Model):
```

Comment:
Please put this in a separate python file.

This class can be extended and it would be a shame for it to end up like [runbot.py](https://github.com/odoo/odoo-extra/blob/master/runbot/runbot.py#L394)